design-proposals/klip-0-add-protobuf-support.md

```diff
@@ -0,0 +1,77 @@
+
+## What is in scope
+
+Adding SerDe support for ProtocolBuffers Version 3 to KSQL.
+
+## What is not in scope
+
+Supporting streams/topics containing multiple ProtocolBuffer schema types.  _See [Design](#Design) section below for more details._
+
+## Value/Return
+
+Out of the box support for running KSQL against Apache Kafka topics containing messages encoded with ProtocolBuffers.
+
+## Public APIS
+
+Proposed extending the `CREATE STREAM` grammar to include a new _value_format_ type **PROTOBUF** as well as adding a new _tableProperty_ named **protobuf_class**.
+
+Example:
+```genericsql
+ksql> CREATE STREAM stream_name (....) WITH (kafka_topic='TopicName', value_format='PROTOBUF', protobuf_class='org.whatever.protobufs.MyProtoBuf$EncodedMessageType');
+```
+
+## Design  
+
+Before digging into the design, it's important to understand how ProtocolBuffers work. A ProtocolBuffer encoded message does **not** include the schema definition. A consumer attempting to deserialize a ProtocolBuffer encoded message must know in advance the schema of that message. Because ProtocolBuffer messages lack this identifying information, services like SchemaRegistry do not exist for ProtocolBuffers.   
+
+**Important:** Because of the limitations in being unable to identify ProtocolBuffer schemas from an encoded message, the proposed design will be limited to only support processing messages from a stream encoded with a single message type/schema.
+
+When defining a new stream, the user would define the following _tableProperties_:
+- value_format: **PROTOBUF**
+- protobuf_class: fully qualified classname for the ProtocolBuffer class represented in the stream's messages.
+
+
+The class defined by the _protobuf_class_ tableProperty would need to already exist at runtime within KSQL's classpath.
+
+When the Serializer/Deserializer instances are created the _protobuf_class_ property would be passed to the KsqlProtobufSerializer/Deserializer via the `public void configure(final Map<String, ?> properties, ...)` method. The Serializer/Deserializer would then be able to instantiate an instance of the configured ProtocolBuffer class as needed. 
+
+The Deserializer `public GenericRow deserialize(final String topic, final byte[] bytes)` implementation would accept a byte array representing a serialized ProtocolBuffer. This method would deserialize the byte array into the configured ProtocolBuffer instance, and apply the appropriate transformation into its equivalent `GenericRow` representation.
+
+The Serializer `public byte[] serialize(final String topic, final GenericRow data)` implementation would accept a GenericRow instance and apply the correct transformation into the appropriate ProtocolBuffer representation. The ProtocolBuffer instance would then be serialized into a byte array and returned.
```

Comment:
What would the serializer do if there were fields in the row that didn't map to fields in the proto class?

Reply:
That's a good question.  How does Avro and JSON handle those scenarios?  Does it just add a new field to the message?

Reply:
@big-andy-coates ping

Reply:
> What would the serializer do if there were fields in the row that didn't map to fields in the proto class?

Have been thinking about this a bit and I personally would go for hard fail. The reason we use protobuf is that we can be very sure that we know exactly what the data schema is. Later on you might introduce flags for different scenarios.

Reply:
@jorgheymans I posted a few weeks ago with no reply, I took the changes here and fleshed them out a bit further.  If you guys are not interested in the changes, no worries, we're running it with success in production for our relatively simple use cases.  The auto schema generation I believe is a major add for ux.

Reply:
@jorgheymans I did not know that, hopefully we can work to get some form of this and mine mainlined.